portabled: add parameters for layered images support

bluca opened the pull request:
Add an --extension parameter to portablectl, and new DBUS methods
to attach/detach/reattach/inspect.
Allows to append separate images on top of the root directory (os-release
will be searched in there) and mount the images using an overlay-like
setup (unit files will be searched in there) using the new ExtensionImages
service option.

Fixes #14697

bluca commented:
RFC because as mentioned in https://github.com/systemd/systemd/issues/14697#issuecomment-670639429 : 

> here's my current thinking on this - now that MountImages has been merged as a primitive, duct-taping it in portabled is easy enough, and in fact I've got it working with a small diff that adds --root and --app-mount-point parameters to portablectl and a new dbus attach method. It works, all is well.
> 
> But - it seems to be it doesn't fit nicely with how images are handled for portabled and nspawn. It doesn't work with discovery and metadata extracting, as only the application image is linked in the portables/machines directory. It feels a bit shoe-horned in.
> 
> So what about the following instead. Establish a parent directory where all application images get mounted in, inside the guest namespace - eg: /run/systemd/app/ (or any other path that fits). And then -
> 
> for single-filesystem images:
> 
> extend /etc/portables and friends to support a drop-in hierarchy structure
> if foo.raw and foo.raw.d/bar.raw are present, foo.raw is RootImage and bar.raw is MountImages on /run/systemd/app/bar/
> same for directory-based images
> os-release is searched in foo.raw, units/nspawn commands are searched for in bar.raw
> for GPT images:
> 
> add a new partition GUID for app partitions, where the label is the name of the application and thus the mount point /run/systemd/app/
> os-release is searched for in the root partition, units/nspawn commands are searched for in the app partition
> What's nice about this scheme is that the next logical step could then be multiple application volumes/services/runners with a single root image, all sharing the same namespaces. Something like pods for podman/k8s.
> 
> Thoughts?

bluca commented:
> F: mksquashfs not found

Don't the bionic images have squashfs-tools installed? How can it be added?

poettering commented:
Hmm, can you elaborate the use case here? i.e. you want to run portable services that are actually the combination of multiple images, i.e. one main image plus a number of additional images mounted in? With an example, how is that supposed to look like in the end?

This PR seems to suggest the main image downloaded is actually going to be one that is mounted into a subdir, instead of the root? What#s the logic behind that? Or am I misunderstanding something?

bluca commented:
> Hmm, can you elaborate the use case here? i.e. you want to run portable services that are actually the combination of multiple images, i.e. one main image plus a number of additional images mounted in? With an example, how is that supposed to look like in the end?
> 
> This PR seems to suggest the main image downloaded is actually going to be one that is mounted into a subdir, instead of the root? What#s the logic behind that? Or am I misunderstanding something?

@poettering  essentially yes - it was a while ago now, but if you remember we talked about this at FOSDEM. The linked RFE has all details, but the summary is that my current use case is similar to what Flatpak/Snaps and others do - a common runtime as the root and each application has its own image on top of that. The use case is the same as other projects implementing the same pattern - I own the runtime env and need to service it faster than application owners can for security updates and so on.

There are different ways to implement this, and there are 3 alternatives I am proposing, in increasing order of complexity and features provided.

The simplest way is what this PR implements - the new --root-image param is where os-release is searched in and it's used with RootImage, and the normal image param is where the units are searched in and it's mounted on the --mount-point path. This is the code as submitted here, minimum required to make this work with no modelling.

This is how it looks like on the cmdline:

portablectl attach --root-image some-distro.raw --mount-point /run/app/foo foo.raw

Which gets foo.service from foo.raw and uses RootImage=some-distro.raw and MountImages=foo.raw:/run/app/foo in the portable.conf generated drop-in.

Then aside from the current implementation, the question I'm asking is whether a stronger modelling is desirable, with a fixed mount point instead of a custom parameterized one, and structuring it in /etc/portables as subdirectories. So the second way, more complex but with a better defined model as a feature, could look like this:

portablectl attach --app foo.raw some-distro.raw

It would result in this under /etc/portables/ :

uuid-some-distro.raw
uuid-some-distro.raw.d/foo.raw

Then as the last pattern with even more features (which I've already been asked for informally) it would be possible to do a podman style "pod", where multiple services run not only with the same runtime/root but also sharing namespaces. It feels like the natural next step for the current set of features.

It would look something like the following:

portablectl attach --app foo.raw --app bar.raw some-distro.raw

It would result in this under /etc/portables/ :

uuid-some-distro.raw
uuid-some-distro.raw.d/foo.raw
uuid-some-distro.raw.d/bar.raw

Then too.service from foo.raw will get RootImage=some-distro.raw and MountImages=foo.raw:/run/systems/portables/foo and same for bar.raw with an extended JoinNamespaces directive.


At the moment, the current implementation as you see here fits my requirements. But working on this I got thinking about the above, and I'd really like to hear feedback about all of this, and what you think makes the most sense here. Thanks!

poettering commented:
hmm, i am not a big fan of overlayfs, but it might be OK as long as things are "mostly ready-only". hence: wouldn't it make sense for RootImage=/path/to/foo.raw to automatically set up overlayfs /path/to/foo.*.overlay.raw onto it?

Or, maybe better: instead of using overlayfs actually "leaf merge" all file systems /path/to/foo.*.extra.raw on top of /path/tp/foo.raw. By "leaf merge" I mean that we'd first mount foo.raw (let's call this one "base"), and then look into foo.*.extra.raw (let's call these ones "extensions") and iterate through all files/dirs in the extensions that are "leaf inodes" in the base. by leaf inodes i mean files or empty dirs, i.e. all inodes that have no further children. 

This would mean, let's say the base contains a typical os tree:

```
/etc/passwd
/etc/group
/usr/bin/ls
/usr/lib64/libc.so.6
/usr/lib/os-release
/opt
/var
…
```

and now extension #1 contains:

```
/opt/myapp/bin/foobar
```

and extension #2 contains:

```
/usr/lib/os-release
```

These three files would then be assembled so that /opt/ from extension #1 would over-mount /opt/ from the base, since /opt prefix of /opt/myapp/bin/foobar that is a leaf. And /usr/lib/os-release would over-mount /usr/lib/os-release from base, since that is the leaf that is the prefix of the same path, if you follow what i mean.

This model has some shortcomings though: everything you add needs to exist as mount point in the base already.

poettering commented:
or hmm, maybe a combination of both models could work, i.e the "leaf merge model" and the "overlayfs model". mount the extensions as simple bind mounts if the deepest existing prefix in the base image is a leaf inode, as already suggested. But if the deepest existing prefix in the base image does not exist, then do an overlayfs mount on the deepest common prefix of both images. And then, declare that any such mounts are strictly read-only, i.e. extensions will only ready-only extend the base, and writable changes may only be made to the base where not obstructed by an extension.

I kinda like that model I must say. It would not require any new directives, but be entirely implicit: you could extend any image like this with more stuff, and the code for it should be reasonably simple still.

bluca commented:
Not a huge fan of overlayfs either, but most importantly there's some conflict with the combination of LSMs I have to use, IPE and SELinux, which means I've got a hard requirement of things working without overlayfs unfortunately. I'll see what I can come up with that uses the same model you are proposing, but without overlay. Should be possible in principle.

poettering commented:
what kind of conflict is that? isn't that something one can fix?

i think overlayfs is actually fine, if used entirely read-only. My problems with it are mostly about using it as writable fs. Hence I think for this usecase it would actually be great.

bluca commented:
> what kind of conflict is that? isn't that something one can fix?
> 
> i think overlayfs is actually fine, if used entirely read-only. My problems with it are mostly about using it as writable fs. Hence I think for this usecase it would actually be great.

If I recall correctly the issue is that due to the way the kernel's tracking of origin of files changes when overlayfs is involved, it is then not possible to enforce code integrity (ie: executables are from a signed verity volume), which is what the IPE LSM is about: https://microsoft.github.io/ipe/
SELinux further complicated things, although I cannot recall how. Sorry, not a kernel dev, and it was discussed long ago, so details are very fuzzy.

I'll spend some time and check if, with your proposal, I'd end up using overlay in my use case or the "leaf merge model" would apply, and if it's the latter then we are good.

bluca commented:
@poettering I've pushed an initial stab at this - not yet full coverage of the multi-image feature, but most of the way there - would like to get some feedback first. Here's what I've done:

1) Added a primitive to compute the list of bind mounts and overlay mounts given two filesystem hierarchies, following the model you recommended
2) Enhanced MountImages: if the destination parameter is not specified, instead of erroring out it now instructs systemd to use the new primitive and do on overlay (this is quite nice and I have other uses for it!)
3) Enhanced portabled/portablectl to take --extra-image as a new parameter, which will end up using the above

It already supports parsing multiple extra images and attaching separate services, but IMHO it makes more sense to support that if we can join the namespaces as well (pod-like feature), so for now it's disabled.

The autodiscovery based on the naming convention is also not implemented, it's next on the list.

I thought about doing the discovery as part of the dissect-image logic rather than one layer above, but in the end I don't like that too much, because I really want to be able to override the default naming and pass custom paths.
Specifically, I need to reuse the same RootImage across many different services, and I need to be able to manage them individually - I don't see a clean way of doing that if the implementation was hidden in the dissect-image functions.

Here's how it looks with some example runs:

```
$ buildx86/portablectl attach --extra-image $PWD/app1.raw $PWD/minimal.raw app1 --runtime
(Matching unit files with prefixes 'app0', 'app1'.)
Bus n/a: changing state UNSET → OPENING
Bus n/a: changing state OPENING → AUTHENTICATING
Bus n/a: changing state AUTHENTICATING → HELLO
Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=Hello cookie=1 reply_cookie=0 signature=n/a error-name=n/a error-message=n/a
Got message type=method_return sender=org.freedesktop.DBus destination=:1.1248 path=n/a interface=n/a member=n/a cookie=1 reply_cookie=1 signature=s error-name=n/a error-message=n/a
Bus n/a: changing state HELLO → RUNNING
Sent message type=method_call sender=n/a destination=org.freedesktop.portable1 path=/org/freedesktop/portable1 interface=org.freedesktop.portable1.Manager member=AttachImageWithExtras cookie=2 reply_cookie=0 signature=sasassbs error-name=n/a error-message=n/a
Got message type=method_return sender=:1.1239 destination=:1.1248 path=n/a interface=n/a member=n/a cookie=21 reply_cookie=2 signature=a(sss) error-name=n/a error-message=n/a
Sent message type=method_call sender=n/a destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=Reload cookie=3 reply_cookie=0 signature=n/a error-name=n/a error-message=n/a
Got message type=method_return sender=:1.1 destination=:1.1248 path=n/a interface=n/a member=n/a cookie=157068 reply_cookie=3 signature=n/a error-name=n/a error-message=n/a
Created directory /run/systemd/system.attached.
Created directory /run/systemd/system.attached/app1.service.d.
Written /run/systemd/system.attached/app1.service.d/20-portable.conf.
Created symlink /run/systemd/system.attached/app1.service.d/10-profile.conf → /usr/lib/systemd/portable/profile/default/service.conf.
Copied /home/bluca/git/systemd/app1.raw → /run/systemd/system.attached/app1.service.
Created symlink /run/portables/app1.raw → /home/bluca/git/systemd/app1.raw.
Bus n/a: changing state RUNNING → CLOSED

$ buildx86/portablectl list
Bus n/a: changing state UNSET → OPENING
Bus n/a: changing state OPENING → AUTHENTICATING
Bus n/a: changing state AUTHENTICATING → HELLO
Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=Hello cookie=1 reply_cookie=0 signature=n/a error-name=n/a error-message=n/a
Got message type=method_return sender=org.freedesktop.DBus destination=:1.1249 path=n/a interface=n/a member=n/a cookie=1 reply_cookie=1 signature=s error-name=n/a error-message=n/a
Bus n/a: changing state HELLO → RUNNING
Sent message type=method_call sender=n/a destination=org.freedesktop.portable1 path=/org/freedesktop/portable1 interface=org.freedesktop.portable1.Manager member=ListImages cookie=2 reply_cookie=0 signature=n/a error-name=n/a error-message=n/a
Got message type=method_return sender=:1.1239 destination=:1.1249 path=n/a interface=n/a member=n/a cookie=23 reply_cookie=2 signature=a(ssbtttso) error-name=n/a error-message=n/a
NAME TYPE RO CRTIME                      MTIME                       USAGE STATE           
app1 raw  no Mon 2020-09-28 13:18:13 BST Mon 2020-09-28 13:18:13 BST 4.0K  attached-runtime

1 images listed.
Bus n/a: changing state RUNNING → CLOSED

$ buildx86/portablectl inspect app1
(Matching unit files with prefix 'app1'.)
Bus n/a: changing state UNSET → OPENING
Bus n/a: changing state OPENING → AUTHENTICATING
Bus n/a: changing state AUTHENTICATING → HELLO
Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=Hello cookie=1 reply_cookie=0 signature=n/a error-name=n/a error-message=n/a
Got message type=method_return sender=org.freedesktop.DBus destination=:1.1263 path=n/a interface=n/a member=n/a cookie=1 reply_cookie=1 signature=s error-name=n/a error-message=n/a
Bus n/a: changing state HELLO → RUNNING
Sent message type=method_call sender=n/a destination=org.freedesktop.portable1 path=/org/freedesktop/portable1 interface=org.freedesktop.portable1.Manager member=GetImageMetadata cookie=2 reply_cookie=0 signature=sas error-name=n/a error-message=n/a
Got message type=method_return sender=:1.1260 destination=:1.1263 path=n/a interface=n/a member=n/a cookie=8 reply_cookie=2 signature=saya{say} error-name=n/a error-message=n/a
Successfully forked off '(pager)' as PID 3427953.
Pager executable is "pager", options "FRSXMK", quit_on_interrupt: yes
Bus n/a: changing state RUNNING → CLOSED
Image:
        /run/portables/app1.raw
Portable Service:
        n/a
Operating System:
        n/a
Unit files:
        app1.service

$ sudo SYSTEMD_LOG_LEVEL=debug buildx86/portablectl detach app1 --runtime
Bus n/a: changing state UNSET → OPENING
Bus n/a: changing state OPENING → AUTHENTICATING
Bus n/a: changing state AUTHENTICATING → HELLO
Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=Hello cookie=1 reply_cookie=0 signature=n/a error-name=n/a error-message=n/a
Got message type=method_return sender=org.freedesktop.DBus destination=:1.1250 path=n/a interface=n/a member=n/a cookie=1 reply_cookie=1 signature=s error-name=n/a error-message=n/a
Bus n/a: changing state HELLO → RUNNING
Sent message type=method_call sender=n/a destination=org.freedesktop.portable1 path=/org/freedesktop/portable1 interface=org.freedesktop.portable1.Manager member=DetachImage cookie=2 reply_cookie=0 signature=sb error-name=n/a error-message=n/a
Got message type=method_return sender=:1.1239 destination=:1.1250 path=n/a interface=n/a member=n/a cookie=26 reply_cookie=2 signature=a(sss) error-name=n/a error-message=n/a
Sent message type=method_call sender=n/a destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=Reload cookie=3 reply_cookie=0 signature=n/a error-name=n/a error-message=n/a
Got message type=method_return sender=:1.1 destination=:1.1250 path=n/a interface=n/a member=n/a cookie=159726 reply_cookie=3 signature=n/a error-name=n/a error-message=n/a
Removed /run/systemd/system.attached/app1.service.
Removed /run/systemd/system.attached/app1.service.d/10-profile.conf.
Removed /run/systemd/system.attached/app1.service.d/20-portable.conf.
Removed /run/systemd/system.attached/app1.service.d.
Removed /run/portables/app1.raw.
Removed /run/systemd/system.attached.
Bus n/a: changing state RUNNING → CLOSED

```

PTAL - thanks!

bluca commented:
Rebased to fix merge conflict.

bluca commented:
Rebased again to fix new merge conflicts with v247. @poettering @keszybz would it be possible to have a review of this new feature as well, please? I applied the suggestions as requested. TIA!

bluca commented:
Rebased to remove merge conflict. Also split out the first commits to implement the overlay suggestion, since I found out good use for that feature in MountImages separately from portabled. Should make it easier to focus on the semantics of portabled, which are trickier to get right.

keszybz commented:
I assume that #18018 should go in first and this doesn't need to be reviewed right now.

bluca commented:
> I assume that #18018 should go in first and this doesn't need to be reviewed right now.

The portabled commit is pretty much independent from the details of the directive used to add the layers, so a review, even high level, of how it's tieing things together would very welcome

keszybz commented:
Sorry for not reviewing this quicker. It needs a rebase now.

bluca commented:
> Sorry for not reviewing this quicker. It needs a rebase now.

That's ok, this is pending the ExtensionImages work anyway

bluca commented:
@poettering @keszybz now that https://github.com/systemd/systemd/pull/18018 is merged (thanks!) this is ready for re-review. It is just a change in portabled to make use of ExtensionImages. PTAL!

keszybz commented:
CI very unhappy.

bluca commented:
> CI very unhappy.

Typo in the DBUS doc that caused a regen to happen - fixed

keszybz commented on man/portablectl.xml:
```diff
@@ -352,6 +352,13 @@
+        <term><option>--extension=</option><replaceable>PATH</replaceable></term>
+
+        <listitem><para>Attach an additional image on top of <replaceable>IMAGE</replaceable>, using an
+        overlay-like setup.</para></listitem>
```
"overlay-like" is unclear... Maybe `Attach an additional image <replaceable>PATH</replaceable> as an overlay on top of <replaceable>IMAGE</replaceable>`?

poettering replied:
should also say that this may be used multiple times on the same command line to attach multiple images. and say in which order the stuff is applied then

bluca replied:
changed - it's IMAGE only for now, pid1 will need ExtensionDirectory= for dirs too (on the TODO list together with nspawn support)

bluca replied:
> should also say that this may be used multiple times on the same command line to attach multiple images. 

actually at the moment multiple --extension results in services being attached separately - I'll fix that (when I implemented it last year I was thinking about future improvement to join namespaces of separate portables)

bluca replied:
done:

```
$ cat /run/systemd/system.attached/app0.service.d/20-portable.conf
# Drop-in created for image '/home/bluca/git/systemd/app0.raw', do not edit.

[Service]
RootImage=/home/bluca/git/systemd/image.raw
Environment=PORTABLE=app0.raw
BindReadOnlyPaths=/etc/os-release:/run/host/os-release
LogExtraFields=PORTABLE=app0.raw
ExtensionImages=/home/bluca/git/systemd/app0.raw
ExtensionImages=/home/bluca/git/systemd/app1.raw
```

keszybz commented on src/portable/portable.c:
```diff
@@ -713,11 +726,15 @@ static int install_chroot_dropin(
                         return -ENOMEM;
+
+                if (m->image && !streq(m->image, image_path))
+                        if (!strextend(&text, "ExtensionImages=", m->image, "\n", NULL))
```
`strextend` does termination internally.

keszybz commented on src/portable/portablectl.c:
```diff
@@ -725,6 +753,12 @@ static int attach_reattach_image(int argc, char *argv[], const char *method) {
         if (r < 0)
                 return bus_log_create_error(r);
 
+        if (streq(method, "AttachImageWithExtensions") || streq(method, "ReattachImageWithExtensions")) {
```
`STR_IN_SET(method, …)`

keszybz commented on src/portable/portablectl.c:
```diff
@@ -733,7 +767,7 @@ static int attach_reattach_image(int argc, char *argv[], const char *method) {
         print_changes(reply);
 
-        if (streq(method, "AttachImage"))
+        if (streq(method, "AttachImage") || streq(method, "AttachImageWithExtensions"))
```
Here too.

keszybz commented on src/portable/portablectl.c:
```diff
@@ -1045,6 +1079,7 @@ static int help(int argc, char *argv[], void *userdata) {
                "     --now                    Immediately start/stop the portable service after\n"
                "                              attach/before detach\n"
                "     --no-block               Don't block waiting for attach --now to complete\n"
+               "     --extension=PATH         Extend the image with overlays\n"
```
'an overlay' ?

Maybe call the argument `IMAGE_OR_DIR` ?

keszybz commented on src/portable/portable.c:
```diff
@@ -980,27 +1016,52 @@ int portable_attach(
                 const char *name_or_path,
                 char **matches,
                 const char *profile,
+                char **extension_images_paths,
```
`extension_image_paths`?

keszybz commented on src/portable/portablectl.c:
```diff
@@ -1185,6 +1222,19 @@ static int parse_argv(int argc, char *argv[]) {
                         break;
 
+                case ARG_EXTENSION: {
+                        _cleanup_free_ char *extension_image = NULL;
```
`_cleanup_` not necessary.

keszybz commented on src/portable/portabled-image-bus.c:
```diff
@@ -238,6 +242,13 @@ int bus_image_common_attach(
         }
 
+        if (sd_bus_message_is_method_call(message, NULL, "AttachImageWithExtensions") ||
+                        sd_bus_message_is_method_call(message, NULL, "AttachWithExtensions")) {
```
indentation

keszybz commented on src/portable/portabled-image-bus.c:
```diff
@@ -246,6 +257,13 @@ int bus_image_common_attach(
                 return r;
 
+        if (sd_bus_message_is_method_call(message, NULL, "AttachImageWithExtensions") ||
+                        sd_bus_message_is_method_call(message, NULL, "AttachWithExtensions")) {
```
Here too.

keszybz commented on src/portable/portabled-image-bus.c:
```diff
@@ -223,12 +225,14 @@ int bus_image_common_attach(
         size_t n_changes = 0;
         int runtime, r;
+        /* Unused for now, but added to the DBUS methods for future-proofing */
+        uint64_t input_flags = 0;
```
I don't think this is necessary.
sd_bus_message_read(3) says:
> Each argument may also be NULL, in which case the value is read and ignored.

But I think that we don't need to read the value to verify message integrity. We can just ignore it, and just rely on the signature of the message being checked. @poettering?

poettering replied:
(that said, I think we should read the param, and refuse if non-null, so that later on clients can check if we support some flag sanely)

bluca replied:
changed to check and reject flags != 0

keszybz commented on src/portable/portabled-image-bus.c:
```diff
@@ -523,6 +543,13 @@ int bus_image_common_reattach(
         }
 
+        if (sd_bus_message_is_method_call(message, NULL, "ReattachImageWithExtensions") ||
+                        sd_bus_message_is_method_call(message, NULL, "ReattachWithExtensions")) {
```
Here and below, every sd_bus_message_* is strangely indented.

keszybz reviewed (commented):
Some superficial comments about style.

poettering commented on src/portable/portable.c:
```diff
@@ -695,12 +708,12 @@ static int install_chroot_dropin(
                 return -ENOMEM;
 
-        text = strjoin(PORTABLE_DROPIN_MARKER_BEGIN, image_path, PORTABLE_DROPIN_MARKER_END "\n");
+        text = strjoin(PORTABLE_DROPIN_MARKER_BEGIN, m->image ?: image_path, PORTABLE_DROPIN_MARKER_END "\n");
```
i don't get this: why is m->image here relevant? i mean, the chroot dropin should always reference the main image, not the extension images. i.e. even if an extension image includes additional service units, which might make sense to honour, they should still be considered to be attached by the main image...

bluca replied:
Because this is used to identify the image to detach/remove, and if it matches on the image then you can't remove a single attached portable service if the main image is shared.

Rather than "main" image, I think it makes more sense to think of these as "root" image (after all, that's what they are used for). In that sense, you then have a 1:N relationship between "root" and "extensions" - multiple extensions (with units) can use the same root image, but an extension (with units) can be used only once since you cannot have duplicated units.

In concrete terms:

```
portablectl attach --extension app0.raw root.raw
portablectl attach --extension app1.raw root.raw
```

This results in two independent portable services being attached and managed separately, using the same rootfs as the base.

```
portablectl detach app0
```

This detaches the first portable service, but not the second. If the root image identified the portable service once attached, both would be removed.

This is the core of my use case - a flatpak-like "runtime" provided by me, used as the rootfs by multiple application images provided by others.

poettering replied:
hmm, but what happens if you have two extensions and two root images, and they are all compatible with each other, and each installs a unit file. And then you attach all combinations: root image a with extension x, root image a with extension y, root image a with extensions x *and* y, and root image a with neither. And then the same for b. now if you only track the extension names in the marker string then detaching a will fuck up b's attachment, and reverse. If you only store the root image name however, you fuck things up differently...

maybe we should always generate an "attachment id" from the images used. the id would be a concat of the root image name, plus all extensions in a well defined order. This would then mean, you'd have to specify --extension= fully again when detaching though, but that would be OK, no?

i.e. when you attach ext1.raw, ext3.raw on top of root.raw, we'd generate the id `root.raw+ext1.raw:ext2.raw` or so from it and include that in the marker.

Wouldn't this make sense?

bluca replied:
That could happen yes, but with the caveat that you'd have to specify different "matches" filters every time, otherwise there would be a clash and the second attachment of the same image would not work.

I'll add the additional ID and push again, hopefully today

bluca replied:
Ok, the marker now is a colon-separated list of all the images including the root one, in the order in which the are used. I've added WithExtensions detach/inspect methods to make use of this. PTAL!

bluca commented on src/shared/discover-image.h:
```diff
@@ -55,6 +55,9 @@ typedef struct Image {
         bool discoverable:1;  /* true if we know for sure that image_find() would find the image given just the short name */
 
+        struct Image **extension_images; /* Will be overlayed on top of the main image */
+        size_t n_extension_images;
```
Sure, changed to a set

bluca commented:
Thanks for the reviews, updated - ptal!

bluca commented:
rebased, integration tests were reshuffled by https://github.com/systemd/systemd/pull/18885

poettering commented on src/portable/portable.c:
```diff
@@ -695,12 +758,21 @@ static int install_chroot_dropin(
 
         if (endswith(m->name, ".service")) {
-                const char *os_release_source;
+                const char *os_release_source, *root_type = IN_SET(type, IMAGE_DIRECTORY, IMAGE_SUBVOLUME) ? "RootDirectory=" : "RootImage=";
```
maybe add a helper call 

```c
const char *root_setting_from_image(Image *i) {
        return IN_SET(type, IMAGE_DIRECTORY, IMAGE_SUBVOLUME) ? "RootDirectory=" : "RootImage=";
}
```

and use that througout? better on the eyes...

poettering commented on src/portable/portable.c:
```diff
@@ -713,11 +785,16 @@ static int install_chroot_dropin(
+                               "Environment=PORTABLE=", basename(m->image ?: image_path), "\n"
                                "BindReadOnlyPaths=", os_release_source, ":/run/host/os-release\n"
-                               "LogExtraFields=PORTABLE=", basename(image_path), "\n"))
+                               "LogExtraFields=PORTABLE=", basename(m->image ?: image_path), "\n"))
```
given the expression in the basename() is getting more complex and complex let's prepare that ahead of time,and reuse it twice

i.e. add 

```c
r = path_extract_filename(m->image ?: image_path, &base_name);
…
```

or so, and then use the extracted value everywhere

poettering commented on src/portable/portable.c:
```diff
@@ -946,32 +1024,53 @@ static int image_symlink(
+                if (symlink(image->path, sl) < 0)
+                        return log_debug_errno(errno, "Failed to link %s %s %s: %m", image->path, special_glyph(SPECIAL_GLYPH_ARROW), sl);
+
+                (void) portable_changes_add(changes, n_changes, PORTABLE_SYMLINK, sl, image->path);
```
instead of repeating basically the same code twice, make this a helper function of its own that you call once for the extensions each and once for the main image from this function?

poettering commented on src/portable/portable.c:
```diff
@@ -980,27 +1079,64 @@ int portable_attach(
+
+                        new = new0(Image, 1);
+                        if (!new)
+                                return -ENOMEM;
```
leaks memory, see comment elsewhere

poettering reviewed (changes requested):
looks pretty good, but a bunch of comments

bluca commented on src/portable/portablectl.c:
```diff
@@ -771,9 +843,53 @@ static int detach_image(int argc, char *argv[], void *userdata) {
+                return bus_log_create_error(r);
+
+        if (!strv_isempty(arg_extension_images)) {
+                r = sd_bus_message_append(m, "t", flags);
```
it's now used for `runtime` as suggested

bluca commented on src/portable/portablectl.c:
```diff
@@ -745,18 +815,20 @@ static int attach_reattach_image(int argc, char *argv[], const char *method) {
         _cleanup_(sd_bus_flush_close_unrefp) sd_bus *bus = NULL;
         _cleanup_free_ char *image = NULL;
+        const char *method;
+        uint64_t flags = 0;
```
refactored as it's used now

bluca commented on man/org.freedesktop.portable1.xml:
```diff
@@ -56,16 +63,37 @@ node /org/freedesktop/portable1 {
+                                in  as extensions,
+                                in  as matches,
+                                in  s profile,
+                                in  b runtime,
```
ok, done for all 3 methods

bluca commented on src/portable/portablectl.c:
```diff
@@ -771,9 +843,53 @@ static int detach_image(int argc, char *argv[], void *userdata) {
+                r = sd_bus_message_close_container(m);
+                if (r < 0)
+                        return bus_log_create_error(r);
+        }
```
done

bluca commented on src/portable/portablectl.c:
```diff
@@ -1185,6 +1304,19 @@ static int parse_argv(int argc, char *argv[]) {
+                        char *extension_image = NULL;
+                        int r;
+
+                        r = parse_path_argument(optarg, false, &extension_image);
```
ok, removed

bluca commented on src/portable/portable.c:
```diff
@@ -401,9 +403,15 @@ static int portable_extract_by_path(
+                        /* When the portable image is layered, the image with units will not
+                         * have a full filesystem, so no os-release - it will be in the root layer */
+                        if (extract_os_release)
+                                flags |= DISSECT_IMAGE_VALIDATE_OS;
```
yes, we do check again - dropped as suggested

bluca commented on src/portable/portable.c:
```diff
@@ -500,7 +518,50 @@ int portable_extract(
+
+                        new = new0(Image, 1);
+                        if (!new)
+                                return -ENOMEM;
```
damn, don't know how I missed this - good catch, fixed both instances

bluca commented on src/portable/portable.c:
```diff
@@ -695,12 +758,21 @@ static int install_chroot_dropin(
+        if (!id)
+                return -ENOMEM;
+        ORDERED_HASHMAP_FOREACH(ext, extension_images)
+                if (!strextend(&id, ":", ext->path))
```
done

bluca commented:
> looks pretty good, but a bunch of comments

Updated and applied all comments, thanks for the review, PTAL!

poettering commented on src/portable/portable.c:
```diff
@@ -74,17 +75,24 @@ static bool unit_match(const char *unit, char **matches) {
+        if (path) {
+                m->image_path = strdup(path);
+                if (!m->image_path)
+                        return NULL;
```
the error path is hosed here. you initialized the object to all zeroes via malloc0(), i.e. `m->fd` is going to be 0, which is a valid fd, and thus if this error path i hit we'l end up closing STDIN by accident. So you either have to initialize m->fd to -1 early on, or you drop the use of `_cleanup_` here for the object, and just insert a regular free() into this error path.

bluca replied:
ouch! good catch, removed _cleanup_

poettering commented on src/portable/portable.c:
```diff
@@ -675,16 +731,53 @@ void portable_changes_free(PortableChange *changes, size_t n_changes) {
+
+        /* If the image is layered, include all layers in the marker as a colon-separated
+         * list of paths, so that we can do exact matches on removal. */
+        text = strjoin(PORTABLE_DROPIN_MARKER_BEGIN, image_path);
```
image_path needs xescape() too

poettering commented on src/portable/portable.c:
```diff
@@ -710,14 +806,23 @@ static int install_chroot_dropin(
+                               "LogExtraFields=PORTABLE=", base_name, "\n"))
                         return -ENOMEM;
+
+                if (m->image_path && !streq(m->image_path, image_path))
```
path_equal()

poettering commented on src/portable/portable.c:
```diff
@@ -975,32 +1081,90 @@ static int install_image_symlink(
         r = image_find_harder(IMAGE_PORTABLE, name_or_path, NULL, &image);
         if (r < 0)
                 return r;
+        if (strv_length(extension_image_paths) > 0) {
```
!strv_is_empty()

poettering commented on src/portable/portable.c:
```diff
@@ -1206,12 +1405,20 @@ int portable_detach(
+                        _cleanup_free_ char *image = NULL;
 
-                        r = set_ensure_consume(&markers, &path_hash_ops_free, TAKE_PTR(marker));
+                        r = extract_first_word(&p, &image, ":", EXTRACT_UNQUOTE|EXTRACT_RETAIN_ESCAPE);
```
should use EXTRACT_UNESCAPE_SEPARATORS i guess.

EXTRACT_UNQUOTE looks wrong

poettering commented on src/portable/portablectl.c:
```diff
@@ -717,14 +769,30 @@ static int attach_reattach_image(int argc, char *argv[], const char *method) {
+        } else {
+                r = sd_bus_message_append(m, "bs", arg_runtime, arg_copy_mode);
+                if (r < 0)
+                        return bus_log_create_error(r);
```
doesn't matter but you could move these two lines after the big if block, since both if branches have the same

poettering commented on src/portable/portablectl.c:
```diff
@@ -771,9 +840,35 @@ static int detach_image(int argc, char *argv[], void *userdata) {
+        } else {
+                r = sd_bus_message_append(m, "b", arg_runtime);
+                if (r < 0)
+                        return bus_log_create_error(r);
```
as above

poettering commented on src/portable/portablectl.c:
```diff
@@ -1185,6 +1283,15 @@ static int parse_argv(int argc, char *argv[]) {
                         break;
 
+                case ARG_EXTENSION: {
+                        int r;
```
"int r;" is a bit special for us, it's that one special variable in our coding style for storing function error codes in. It hence gets the special privilege to always be declared at the top-level scope of the function. Let's do that here too, move the "int r;" to the top of the function. This then allows you to get rid of the {}

poettering commented on src/portable/portabled-bus.c:
```diff
@@ -265,10 +267,39 @@ static int method_detach_image(sd_bus_message *message, void *userdata, sd_bus_e
+                        return r;
+        }
+
+        if (sd_bus_message_is_method_call(message, NULL, "DetachImageWithExtensions")) {
```
seems you can merge this with the previous if block

bluca commented:
@poettering thank you, updated again - ptal!

poettering commented:
lgtm
